x/wasm/keeper/keeper.go

```diff
@@ -1161,30 +1144,38 @@ func (c BankCoinTransferrer) TransferCoins(parentCtx sdk.Context, fromAddr sdk.A
 	return nil
 }
 
-var _ CoinPruner = CoinBurner{}
+var _ CoinPruner = VestingCoinBurner{}
 
-// CoinBurner default implementation for CoinPruner to burn the coins
-type CoinBurner struct {
+// VestingCoinBurner default implementation for CoinPruner to burn the coins
+type VestingCoinBurner struct {
 	bank types.BankKeeper
 }
 
-// NewCoinBurner constructor
-func NewCoinBurner(bank types.BankKeeper) CoinBurner {
+// NewVestingCoinBurner constructor
+func NewVestingCoinBurner(bank types.BankKeeper) VestingCoinBurner {
 	if bank == nil {
 		panic("bank keeper must not be nil")
 	}
-	return CoinBurner{bank: bank}
+	return VestingCoinBurner{bank: bank}
 }
 
 // PruneBalances burns all coins owned by the account.
-func (b CoinBurner) PruneBalances(ctx sdk.Context, address sdk.AccAddress) error {
-	if amt := b.bank.GetAllBalances(ctx, address); !amt.IsZero() {
-		if err := b.bank.SendCoinsFromAccountToModule(ctx, address, types.ModuleName, amt); err != nil {
-			return sdkerrors.Wrap(err, "prune account balance")
-		}
-		if err := b.bank.BurnCoins(ctx, types.ModuleName, amt); err != nil {
-			return sdkerrors.Wrap(err, "burn account balance")
-		}
+func (b VestingCoinBurner) PruneBalances(ctx sdk.Context, existingAcc authtypes.AccountI) error {
+	v, ok := existingAcc.(vestingexported.VestingAccount)
+	if !ok {
```

Comment:
This was built with the assumption that also other than SDK account types can exists in a chain. The `CoinPruner.PruneBalances` method is an extension point that can be used so that chains can handle their account types and/or vesting account types as they want. Another implementation could also be sending the tokens to the community pool or an escrow (not saying this makes sense).  
The concrete struct type is `VestingCoinBurner`. I thought this verbose enough and explicit addressing the SDK vesting account types only (or return error).
I can add some more code doc to this but switching to VestingAccount types only would not work due to the custom account types. WDYT?

Reply:
Ah, okay, so this PruneBalances is not on the Keeper, but can easily be set by an external app. Now I understand the design.

> Throwing types.ErrAccountExists though would better fit one level where it was before. I will modify the method signature to return a handled bool for this

No need for `handled` unless you prefer it as well, I think returning `types.ErrAccountExists` is nice

Reply:
You convinced me that we need to keep the logic here. That does make sense.

However, the name is misleading. `PruneBalances` is what you do when you encounter a vesting account.

Maybe `CleanupExistingAccount` would be a better name (as is more general and fits what you want as an extension point). And if it is a vesting account, you `PruneBalances`. Maybe other accounts have different cleanup methods? And if there is no cleanup possible (existing module account), returning some `types.ErrCannotReuseAccount` error or such sounds good.

Reply:
👍 `CleanupExistingAccount` makes sense as a more general term. I prefer the `handled` result value, too as it brings back the control flow to the caller.